src/main/java/org/entur/lamassu/controller/GBFSV2FeedController.java

```diff
@@ -98,10 +104,47 @@ public ResponseEntity<Object> getGbfsFeedForProvider(
     } catch (IllegalArgumentException e) {
       throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
     } catch (NoSuchElementException e) {
+      // if system_id is well known, feed could exist according to the spec but
+      // is not available (e.g. due to upstream issues), we respond with 502 and not 4xx.
+      // In case the feed is initialised (gbfs feed is available),
+      // we throw 502 if the feed should exist according to the gbfs, else 404.
+      if (feedProviderService.getFeedProviderBySystemId(systemId) != null) {
+        if (mightOrShouldFeedExist(systemId, feed)) {
+          throw new ResponseStatusException(HttpStatus.BAD_GATEWAY);
+        }
+        throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+      }
       throw new ResponseStatusException(HttpStatus.NOT_FOUND);
     }
   }
 
+  /*
+    Returns true, if either the discoveryFile (gbf file) is not yet cached,
+    or the requested feed is published in the discovery file.
+   */
+  protected boolean mightOrShouldFeedExist(String systemId, String feed) {
```

Comment:
One alternative to returning a boolean here which you have to check at the call site to decide which exception to throw, is to throw the exception directly from within this method. I think that's ok since it's private / local. Could be renamed to "throwIfShouldExist" or something similar.

Reply:
Yes, and I'll move the checking directly into getFeed and throwing the intended exception right there, instead of checking it in the exception handling again.